### PR TITLE
Azure Event Hubs pubsub component

### PR DESCRIPTION
**Pull request by abhirockzz**
# Description

Azure Event Hubs as a pubsub component

## Issue reference

Issue #241 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

**Comment by yaron2** [comment]
This is a great and welcome addition

**Comment by abhirockzz** [comment]
Please see the [original issue for a summary](https://github.com/dapr/components-contrib/issues/241#issuecomment-593790669) of how to use this pubsub component

**Comment by yaron2** [comment]
How did you test this component?

**Comment by abhirockzz** [comment]
> How did you test this component?

Just like any other - integrated this into dapr core ([much easier with this reference!](https://github.com/dapr/components-contrib/blob/master/docs/developing-component.md)) and ran tests with an actual event hubs namespace + topic.

**Comment by abhirockzz** [comment]
I can reproduce the `check-diff` error on my local setup as well

```
abhishekgupta$ make go.mod check-diff
go mod tidy
git diff --exit-code ./go.mod # check no changes
diff --git a/go.mod b/go.mod
index 13d07e2..a4bc86d 100644
--- a/go.mod
+++ b/go.mod
@@ -60,7 +60,7 @@ require (
        go.etcd.io/etcd v3.3.17+incompatible
        go.mongodb.org/mongo-driver v1.1.2
        go.opencensus.io v0.22.3
-       golang.org/x/crypto v0.0.0-20191206172530-e9b2fee46413
+       golang.org/x/crypto v0.0.0-20200206161412-a0c6ece9d31a
        golang.org/x/net v0.0.0-20200202094626-16171245cfb2
        golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d
        google.golang.org/api v0.15.0
make: *** [check-diff] Error 1
```

**Comment by yaron2** [comment]
> I can reproduce the `check-diff` error on my local setup as well
> 
> ```
> abhishekgupta$ make go.mod check-diff
> go mod tidy
> git diff --exit-code ./go.mod # check no changes
> diff --git a/go.mod b/go.mod
> index 13d07e2..a4bc86d 100644
> --- a/go.mod
> +++ b/go.mod
> @@ -60,7 +60,7 @@ require (
>         go.etcd.io/etcd v3.3.17+incompatible
>         go.mongodb.org/mongo-driver v1.1.2
>         go.opencensus.io v0.22.3
> -       golang.org/x/crypto v0.0.0-20191206172530-e9b2fee46413
> +       golang.org/x/crypto v0.0.0-20200206161412-a0c6ece9d31a
>         golang.org/x/net v0.0.0-20200202094626-16171245cfb2
>         golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d
>         google.golang.org/api v0.15.0
> make: *** [check-diff] Error 1
> ```

Which version of Go are you using? make sure you're with 1.13.x

**Comment by yaron2** [comment]
@youngbupark can you please review this?

**Comment by abhirockzz** [comment]
> > I can reproduce the `check-diff` error on my local setup as well
> > ```
> > abhishekgupta$ make go.mod check-diff
> > go mod tidy
> > git diff --exit-code ./go.mod # check no changes
> > diff --git a/go.mod b/go.mod
> > index 13d07e2..a4bc86d 100644
> > --- a/go.mod
> > +++ b/go.mod
> > @@ -60,7 +60,7 @@ require (
> >         go.etcd.io/etcd v3.3.17+incompatible
> >         go.mongodb.org/mongo-driver v1.1.2
> >         go.opencensus.io v0.22.3
> > -       golang.org/x/crypto v0.0.0-20191206172530-e9b2fee46413
> > +       golang.org/x/crypto v0.0.0-20200206161412-a0c6ece9d31a
> >         golang.org/x/net v0.0.0-20200202094626-16171245cfb2
> >         golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d
> >         google.golang.org/api v0.15.0
> > make: *** [check-diff] Error 1
> > ```
> 
> Which version of Go are you using? make sure you're with 1.13.x

Yes I was on 1.13.x earlier. Upgraded to 1.14 and see the same issue

**Comment by yaron2** [comment]
Switch to 1.13 and run go mod tidy

**Comment by abhirockzz** [comment]
> Switch to 1.13 and run go mod tidy

Tried that. `go mod tidy` runs fine, but the `git diff --exit-code ./go.mod` returns the same error

**Comment by yaron2** [comment]
@abhirockzz I took your branch locally, and `make go.mod` finishes with no errors..

File: pubsub/azure/eventhubs/eventhubs.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	"github.com/Azure/azure-event-hubs-go/storage"
+	"github.com/Azure/azure-storage-blob-go/azblob"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/dapr/components-contrib/pubsub"
+	"github.com/dapr/dapr/pkg/logger"
+)
+
+const (
+	// metadata
+
```
Comment (youngbupark) [inline pubsub/azure/eventhubs/eventhubs.go]: nit: unnecessary return

File: pubsub/azure/eventhubs/eventhubs.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	connectionString = "connectionString"
+	consumerID       = "consumerID" // passed by dapr runtime
+	// required by subscriber
+	storageAccountName   = "storageAccountName"
+	storageAccountKey    = "storageAccountKey"
+	storageContainerName = "storageContainerName"
+
+	// errors
+
```
Comment (youngbupark) [inline pubsub/azure/eventhubs/eventhubs.go]: nit: unnecessary return

File: pubsub/azure/eventhubs/eventhubs.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	err := aeh.hub.Send(context.Background(), &eventhub.Event{Data: req.Data})
+	if err != nil {
+		return fmt.Errorf("error from publish: %s", err)
+	}
+	return nil
+}
+
+// Subscribe recieves data from Azure Event Hubs
+func (aeh *AzureEventHubs) Subscribe(req pubsub.SubscribeRequest, handler func(msg *pubsub.NewMessage) error) error {
+
```
Comment (youngbupark) [inline pubsub/azure/eventhubs/eventhubs.go]: nit: unnecessary return

File: pubsub/azure/eventhubs/eventhubs.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+		return err
+	}
+
+	leaserCheckpointer, err := storage.NewStorageLeaserCheckpointer(cred, aeh.metadata.storageAccountName, aeh.metadata.storageContainerName, azure.PublicCloud)
+	if err != nil {
+		return err
+	}
+
+	processor, err := eph.NewFromConnectionString(context.Background(), aeh.metadata.connectionString, leaserCheckpointer, leaserCheckpointer, eph.WithNoBanner(), eph.WithConsumerGroup(aeh.metadata.consumerGroup))
+
```
Comment (youngbupark) [inline pubsub/azure/eventhubs/eventhubs.go]: nit: unnecessary return

File: pubsub/azure/eventhubs/eventhubs.go (hunk 1 above)
Comment (abhirockzz) [reply pubsub/azure/eventhubs/eventhubs.go]: I did not quite understand what "unnecessary return" means in this context (did you mean "space"?). I'd be happy to make changes if you can help clarify. Thanks!

**Comment by abhirockzz** [comment]
> LGTM except for lint :) after removing spaces, please run make lint.
> 
> Thanks for your contribution :)

Sure. `make lint` often changes other (unrelated) files and they end up being included in the PR. IS that ok?

**Comment by youngbupark** [comment]
> > LGTM except for lint :) after removing spaces, please run make lint.
> > Thanks for your contribution :)
> 
> Sure. `make lint` often changes other (unrelated) files and they end up being included in the PR. IS that ok?

I created PR - https://github.com/dapr/components-contrib/pull/257

**Comment by youngbupark** [comment]
> > > LGTM except for lint :) after removing spaces, please run make lint.
> > > Thanks for your contribution :)
> > 
> > 
> > Sure. `make lint` often changes other (unrelated) files and they end up being included in the PR. IS that ok?
> 
> I created PR - #257

now master is clean :)

**Comment by yaron2** [comment]
I will merge this now, but this will only come out in a version after `0.5.0` for Dapr.

**Pull request merged**